File: hk/hackathon/frontend-react/src/engines/sla.js
// Port of backend/app/engines/sla.py — same guards, not refactored.

export function simpleForecast(history) {
  const n = history.length;
  if (n < 3) return history[n - 1] ?? 0;
  const xs = Array.from({ length: n }, (_, i) => i);
  const meanX = xs.reduce((a, b) => a + b, 0) / n;
  const meanY = history.reduce((a, b) => a + b, 0) / n;
  let num = 0, den = 0;
  for (let i = 0; i < n; i++) {
    num += (xs[i] - meanX) * (history[i] - meanY);
    den += (xs[i] - meanX) ** 2;
  }
  const slope = num / (den || 1e-6);
  return history[n - 1] + slope * 3; // project 3 ticks ahead
}

export function evaluateSla(latencyMs, limitMs, predictedLatencyMs = null) {
  if (latencyMs > limitMs) return { state: "VIOLATION", risk: 1.0 };
  if (predictedLatencyMs !== null) {
    const projectedBreach = predictedLatencyMs > limitMs * 1.05;
    const material = latencyMs > limitMs * 0.30;
    if (projectedBreach && material) {
      const risk = Math.min(1.0, (predictedLatencyMs - limitMs) / limitMs + 0.5);
      return { state: "AT_RISK", risk: Number(risk.toFixed(3)) };
    }
  }
  const risk = Number(Math.min(0.4, latencyMs / limitMs).toFixed(3));
  return { state: "HEALTHY", risk };
}

File: hk/hackathon/backend/app/engines/sla.py
"""SLA state machine: HEALTHY / AT_RISK / VIOLATION.
AT_RISK is reachable from a forecast alone, gated by two guards so a cold
baseline can't be extrapolated into false alarms.
"""


def evaluate_sla(latency_ms: float, limit_ms: float, predicted_latency_ms: float | None = None):
    if latency_ms > limit_ms:
        return "VIOLATION", 1.0
    if predicted_latency_ms is not None:
        projected_breach = predicted_latency_ms > limit_ms * 1.05
        material = latency_ms > limit_ms * 0.30
        if projected_breach and material:
            risk = min(1.0, (predicted_latency_ms - limit_ms) / limit_ms + 0.5)
            return "AT_RISK", round(risk, 3)
    risk = round(min(0.4, latency_ms / limit_ms), 3)
    return "HEALTHY", risk


def simple_forecast(history: list[float]) -> float:
    """OLS trend over the last N samples -> next-tick projection."""
    n = len(history)
    if n < 3:
        return history[-1] if history else 0.0
    xs = list(range(n))
    mean_x = sum(xs) / n
    mean_y = sum(history) / n
    num = sum((xs[i] - mean_x) * (history[i] - mean_y) for i in range(n))
    den = sum((xs[i] - mean_x) ** 2 for i in range(n)) or 1e-6
    slope = num / den
    return history[-1] + slope * 3  # project 3 ticks ahead
